File: tests/oracle/rule_identityel_opcodeset_1204.cc
/*
 * RULE-IDENTITYEL-OPCODESET-0001
 *
 * Locked Ghidra 12.0.4 oracle for RuleIdentityEl's public Rule/ActionPool
 * contract (ruleaction.hh:673-682, ruleaction.cc:3668-3702,
 * action.cc:740-750/822-887).  A production ActionPool containing only the
 * production RuleIdentityEl is driven to a fixed point over one ordered bank
 * of synthetic P-code operations.  No expected result is encoded here: the
 * runner records and compares the complete observations from both engines.
 */

#include "bfd_arch.hh"
#include "libdecomp.hh"
#include "ruleaction.hh"

#include <iostream>
#include <iterator>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::ostringstream;
using std::string;
using std::vector;

struct CaseRecord {
  string name;
  PcodeOp *op;
  Varnode *lhs;
  Varnode *rhs;
  Varnode *output;
};

AddrSpace *codeSpace;
AddrSpace *regSpace;
uintb nextAddress = 0x6100;
uintb nextRegister = 0x400;

Varnode *newInput(Funcdata &fd,int4 size)
{
  Varnode *vn = fd.newVarnode(size,regSpace,nextRegister);
  nextRegister += 0x10;
  return fd.setInputVarnode(vn);
}

CaseRecord makeCase(Funcdata &fd,BlockBasic *block,const string &name,
                    OpCode opcode,Varnode *lhs,Varnode *rhs,int4 outputSize)
{
  PcodeOp *op = fd.newOp(2,Address(codeSpace,nextAddress++));
  fd.opSetOpcode(op,opcode);
  fd.opSetInput(op,lhs,0);
  fd.opSetInput(op,rhs,1);
  Varnode *output = fd.newUniqueOut(outputSize,op);
  fd.opInsertEnd(op,block);
  CaseRecord result = { name,op,lhs,rhs,output };
  return result;
}

string caseName(const PcodeOp *op,const vector<CaseRecord> &cases)
{
  for(size_t i=0;i<cases.size();++i)
    if (cases[i].op == op) return cases[i].name;
  return "?";
}

string descendOrder(const Varnode *vn,const vector<CaseRecord> &cases)
{
  ostringstream out;
  bool first = true;
  for(list<PcodeOp *>::const_iterator iter=vn->beginDescend();
      iter!=vn->endDescend();++iter) {
    if (!first) out << ',';
    first = false;
    out << caseName(*iter,cases);
  }
  return out.str();
}

string blockOrder(const BlockBasic *block,const vector<CaseRecord> &cases)
{
  ostringstream out;
  bool first = true;
  for(list<PcodeOp *>::const_iterator iter=block->beginOp();
      iter!=block->endOp();++iter) {
    if (!first) out << ',';
    first = false;
    out << caseName(*iter,cases);
  }
  return out.str();
}

string slotIdentity(const PcodeOp *op,int4 slot,const CaseRecord &record)
{
  if (slot >= op->numInput()) return "-";
  const Varnode *vn = op->getIn(slot);
  if (vn == record.lhs) return "lhs";
  if (vn == record.rhs) return "rhs";
  return "other";
}

void emitCase(const CaseRecord &record,const BlockBasic *block)
{
  PcodeOp *op = record.op;
  std::cout << "case=" << record.name
            << "|opcode=" << static_cast<int4>(op->code())
            << "|inputs=" << op->numInput()
            << "|slot0=" << slotIdentity(op,0,record)
            << "|slot1=" << slotIdentity(op,1,record)
            << "|output_same=" << (op->getOut()==record.output ? 1 : 0)
            << "|output_def=" << (record.output->getDef()==op ? 1 : 0)
            << "|rhs_desc="
            << std::distance(record.rhs->beginDescend(),record.rhs->endDescend())
            << "|parent_same=" << (op->getParent()==block ? 1 : 0)
            << "|dead=" << (op->isDead() ? 1 : 0)
            << '\n';
}

void emitOpList(Rule &rule)
{
  vector<uint4> oplist;
  rule.getOpList(oplist);
  std::cout << "oplist=";
  for(size_t i=0;i<oplist.size();++i) {
    if (i != 0) std::cout << ',';
    std::cout << oplist[i];
  }
  std::cout << "|count=" << oplist.size() << '\n';
}

void run(Funcdata &fd)
{
  BlockGraph &graph = const_cast<BlockGraph &>(fd.getBasicBlocks());
  BlockBasic *block = graph.newBlockBasic(&fd);
  Varnode *x4 = newInput(fd,4);
  Varnode *b1 = newInput(fd,1);
  vector<CaseRecord> cases;

  cases.push_back(makeCase(fd,block,"int_add_zero",CPUI_INT_ADD,x4,
                           fd.newConstant(4,0),4));
  cases.push_back(makeCase(fd,block,"int_xor_zero",CPUI_INT_XOR,x4,
                           fd.newConstant(4,0),4));
  cases.push_back(makeCase(fd,block,"int_or_zero",CPUI_INT_OR,x4,
                           fd.newConstant(4,0),4));
  cases.push_back(makeCase(fd,block,"bool_xor_zero",CPUI_BOOL_XOR,b1,
                           fd.newConstant(1,0),1));
  cases.push_back(makeCase(fd,block,"bool_or_zero",CPUI_BOOL_OR,b1,
                           fd.newConstant(1,0),1));
  cases.push_back(makeCase(fd,block,"mult_zero",CPUI_INT_MULT,x4,
                           fd.newConstant(4,0),4));
  cases.push_back(makeCase(fd,block,"mult_one",CPUI_INT_MULT,x4,
                           fd.newConstant(4,1),4));
  cases.push_back(makeCase(fd,block,"mult_two_guard",CPUI_INT_MULT,x4,
                           fd.newConstant(4,2),4));
  cases.push_back(makeCase(fd,block,"int_sub_zero_dispatch_neg",CPUI_INT_SUB,x4,
                           fd.newConstant(4,0),4));
  cases.push_back(makeCase(fd,block,"int_and_zero_dispatch_neg",CPUI_INT_AND,x4,
                           fd.newConstant(4,0),4));
  cases.push_back(makeCase(fd,block,"int_add_nonconst_guard",CPUI_INT_ADD,x4,
                           newInput(fd,4),4));

  RuleIdentityEl *rule = new RuleIdentityEl("analysis");
  ActionPool pool(Action::rule_repeatapply,"identity_pool");
  pool.addRule(rule);

  emitOpList(*rule);
  std::cout << "lookup=production"
            << "|new=" << (pool.getSubRule("identityel")==rule ? 1 : 0)
            << "|old=" << (pool.getSubRule("identity_el")==0 ? 0 : 1)
            << '\n';
  std::cout << "rule_name=" << rule->getName() << '\n';
  std::cout << "desc=x4|stage=before|ops=" << descendOrder(x4,cases) << '\n';
  std::cout << "desc=b1|stage=before|ops=" << descendOrder(b1,cases) << '\n';
  std::cout << "block|stage=before|ops=" << blockOrder(block,cases) << '\n';

  int4 result = pool.perform(fd);
  std::cout << "pool|perform=" << result
            << "|status=" << pool.getStatus()
            << "|tests=" << pool.getNumTests()
            << "|apply=" << pool.getNumApply()
            << "|rule_tests=" << rule->getNumTests()
            << "|rule_apply=" << rule->getNumApply()
            << '\n';
  for(size_t i=0;i<cases.size();++i)
    emitCase(cases[i],block);
  std::cout << "desc=x4|stage=after|ops=" << descendOrder(x4,cases) << '\n';
  std::cout << "desc=b1|stage=after|ops=" << descendOrder(b1,cases) << '\n';
  std::cout << "block|stage=after|ops=" << blockOrder(block,cases) << '\n';
}

} // anonymous namespace

int main(int argc,char **argv)
{
  if (argc != 3) {
    std::cerr << "usage: rule_identityel_opcodeset_1204 SPEC_ROOT CURL_BINARY\n";
    return 2;
  }
  try {
    vector<string> specPaths;
    specPaths.push_back(argv[1]);
    startDecompilerLibrary(specPaths);
    {
      std::ostringstream diagnostics;
      BfdArchitecture architecture(argv[2],"default",&diagnostics);
      DocumentStorage store;
      architecture.init(store);
      architecture.readLoaderSymbols("::");
      Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("GetStr");
      if (fd == (Funcdata *)0)
        throw std::runtime_error("GetStr was not found in the BFD symbol table");
      if (fd->getName() != "GetStr" ||
          fd->getAddress().getOffset() != 0x36d0 || fd->getSize() != 0)
        throw std::runtime_error("GetStr input identity drifted");
      if (architecture.archid != "x86:LE:64:default:gcc")
        throw std::runtime_error("runtime architecture/compiler drifted: " +
                                 architecture.archid);
      codeSpace = architecture.getDefaultCodeSpace();
      regSpace = architecture.getSpaceByName("register");
      if (regSpace == (AddrSpace *)0)
        throw std::runtime_error("fixture requires register space");
      run(*fd);
    }
    shutdownDecompilerLibrary();
    return 0;
  }
  catch(const ghidra::LowlevelError &error) {
    std::cerr << "Ghidra LowlevelError: " << error.explain << '\n';
  }
  catch(const ghidra::DecoderError &error) {
    std::cerr << "Ghidra DecoderError: " << error.explain << '\n';
  }
  catch(const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
  }
  return 1;
}

File: tests/oracle/rule_identityel_opcodeset_1204.rs
//! RULE-IDENTITYEL-OPCODESET-0001: Rust twin of the locked Ghidra 12.0.4
//! RuleIdentityEl fixture.
//!
//! The production RuleIdentityEl is registered through the production
//! ActionPool and driven through Action::perform. Fixture-only counters wrap
//! the Rule without changing its opcode list or mutation behavior. Rust's
//! production ActionPool currently lacks Ghidra's getSubRule API; the lookup
//! record makes that known whole-observation mismatch explicit while the
//! remaining records are compared byte-for-byte with the oracle.

use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::{Arc, RwLock};

use rugra::action::{Action, ActionPool, ActionState, Rule};
use rugra::address::Address;
use rugra::arch::Architecture;
use rugra::block::FlowBlock;
use rugra::funcdata::Funcdata;
use rugra::op::PcodeOpRef;
use rugra::opcodes::OpCode;
use rugra::ruleaction::RuleIdentityEl;
use rugra::space::AddressSpace;
use rugra::varnode::Varnode;

type VarnodeRef = Arc<RwLock<Varnode>>;

struct CaseRecord {
    name: &'static str,
    op: PcodeOpRef,
    lhs: VarnodeRef,
    rhs: VarnodeRef,
    output: VarnodeRef,
}

#[derive(Default)]
struct RuleCounters {
    tests: AtomicU32,
    apply: AtomicU32,
}

struct CountingIdentityEl {
    inner: RuleIdentityEl,
    counters: Arc<RuleCounters>,
}

impl Rule for CountingIdentityEl {
    fn apply_op(
        &self,
        op: &Arc<RwLock<rugra::op::PcodeOp>>,
        fd: &mut Funcdata,
    ) -> rugra::Result<i32> {
        self.counters.tests.fetch_add(1, Ordering::SeqCst);
        let result = self.inner.apply_op(op, fd)?;
        if result > 0 {
            self.counters.apply.fetch_add(1, Ordering::SeqCst);
        }
        Ok(result)
    }

    fn get_name(&self) -> &str {
        self.inner.get_name()
    }

    fn get_opcodes(&self) -> Vec<OpCode> {
        self.inner.get_opcodes()
    }
}

fn new_input(fd: &mut Funcdata, size: usize, next_register: &mut u64) -> VarnodeRef {
    let vn = fd
        .vbank
        .create_with_space(size, AddressSpace::Register, *next_register);
    *next_register += 0x10;
    fd.set_input_varnode(vn)
}

fn make_case(
    fd: &mut Funcdata,
    block: &Arc<RwLock<dyn FlowBlock + Send + Sync>>,
    next_address: &mut u64,
    name: &'static str,
    opcode: OpCode,
    lhs: VarnodeRef,
    rhs: VarnodeRef,
    output_size: usize,
) -> CaseRecord {
    let op = fd.new_op(2, Address::new(*next_address));
    *next_address += 1;
    fd.op_set_opcode(&op, opcode);
    fd.op_set_input(&op, lhs.clone(), 0);
    fd.op_set_input(&op, rhs.clone(), 1);
    let output = fd.new_unique_out(output_size, &op);
    fd.op_insert_end(&op, block);
    CaseRecord {
        name,
        op,
        lhs,
        rhs,
        output,
    }
}

fn case_name(op: &Arc<RwLock<rugra::op::PcodeOp>>, cases: &[CaseRecord]) -> &'static str {
    cases
        .iter()
        .find(|record| Arc::ptr_eq(&record.op.0, op))
        .map_or("?", |record| record.name)
}

fn descend_order(vn: &VarnodeRef, cases: &[CaseRecord]) -> String {
    vn.read()
        .unwrap()
        .descend_iter()
        .map(|op| case_name(&op, cases))
        .collect::<Vec<_>>()
        .join(",")
}

fn block_order(block: &Arc<RwLock<dyn FlowBlock + Send + Sync>>, cases: &[CaseRecord]) -> String {
    block
        .read()
        .unwrap()
        .get_ops()
        .iter()
        .map(|op| case_name(&op.0, cases))
        .collect::<Vec<_>>()
        .join(",")
}

fn slot_identity(op: &rugra::op::PcodeOp, slot: usize, record: &CaseRecord) -> &'static str {
    let Some(vn) = op.get_in(slot) else {
        return "-";
    };
    if Arc::ptr_eq(vn, &record.lhs) {
        "lhs"
    } else if Arc::ptr_eq(vn, &record.rhs) {
        "rhs"
    } else {
        "other"
    }
}

fn emit_case(record: &CaseRecord, block: &Arc<RwLock<dyn FlowBlock + Send + Sync>>) {
    let op = record.op.0.read().unwrap();
    let output_def = record
        .output
        .read()
        .unwrap()
        .get_def()
        .is_some_and(|definition| Arc::ptr_eq(&definition, &record.op.0));
    let parent_same = op
        .parent
        .as_ref()
        .and_then(std::sync::Weak::upgrade)
        .is_some_and(|parent| Arc::ptr_eq(&parent, block));
    println!(
        "case={}|opcode={}|inputs={}|slot0={}|slot1={}|output_same={}|output_def={}|rhs_desc={}|parent_same={}|dead={}",
        record.name,
        op.opcode as i32,
        op.num_input(),
        slot_identity(&op, 0, record),
        slot_identity(&op, 1, record),
        usize::from(
            op.output
                .as_ref()
                .is_some_and(|output| Arc::ptr_eq(output, &record.output)),
        ),
        usize::from(output_def),
        record.rhs.read().unwrap().count_descends(),
        usize::from(parent_same),
        usize::from(op.is_dead()),
    );
}

fn run() {
    let mut fd = Funcdata::new("GetStr", Address::new(0x36d0), 0);
    let mut architecture = Architecture::new();
    architecture.archid = "x86:LE:64:default:gcc".to_string();
    fd.set_arch(Arc::new(architecture));
    let block = fd.create_new_block();
    let mut next_address = 0x6100;
    let mut next_register = 0x400;
    let x4 = new_input(&mut fd, 4, &mut next_register);
    let b1 = new_input(&mut fd, 1, &mut next_register);
    let mut cases = Vec::new();

    let rhs = fd.new_constant(4, 0);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "int_add_zero",
        OpCode::CPUI_INT_ADD,
        x4.clone(),
        rhs,
        4,
    ));
    let rhs = fd.new_constant(4, 0);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "int_xor_zero",
        OpCode::CPUI_INT_XOR,
        x4.clone(),
        rhs,
        4,
    ));
    let rhs = fd.new_constant(4, 0);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "int_or_zero",
        OpCode::CPUI_INT_OR,
        x4.clone(),
        rhs,
        4,
    ));
    let rhs = fd.new_constant(1, 0);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "bool_xor_zero",
        OpCode::CPUI_BOOL_XOR,
        b1.clone(),
        rhs,
        1,
    ));
    let rhs = fd.new_constant(1, 0);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "bool_or_zero",
        OpCode::CPUI_BOOL_OR,
        b1.clone(),
        rhs,
        1,
    ));
    let rhs = fd.new_constant(4, 0);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "mult_zero",
        OpCode::CPUI_INT_MULT,
        x4.clone(),
        rhs,
        4,
    ));
    let rhs = fd.new_constant(4, 1);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "mult_one",
        OpCode::CPUI_INT_MULT,
        x4.clone(),
        rhs,
        4,
    ));
    let rhs = fd.new_constant(4, 2);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "mult_two_guard",
        OpCode::CPUI_INT_MULT,
        x4.clone(),
        rhs,
        4,
    ));
    let rhs = fd.new_constant(4, 0);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "int_sub_zero_dispatch_neg",
        OpCode::CPUI_INT_SUB,
        x4.clone(),
        rhs,
        4,
    ));
    let rhs = fd.new_constant(4, 0);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "int_and_zero_dispatch_neg",
        OpCode::CPUI_INT_AND,
        x4.clone(),
        rhs,
        4,
    ));
    let rhs = new_input(&mut fd, 4, &mut next_register);
    cases.push(make_case(
        &mut fd,
        &block,
        &mut next_address,
        "int_add_nonconst_guard",
        OpCode::CPUI_INT_ADD,
        x4.clone(),
        rhs,
        4,
    ));

    let counters = Arc::new(RuleCounters::default());
    let mut pool = ActionPool::new("identity_pool");
    pool.add_rule(Box::new(CountingIdentityEl {
        inner: RuleIdentityEl::new(),
        counters: counters.clone(),
    }));

    let opcodes = RuleIdentityEl::new().get_opcodes();
    println!(
        "oplist={}|count={}",
        opcodes
            .iter()
            .map(|opcode| format!("{}", *opcode as i32))
            .collect::<Vec<_>>()
            .join(","),
        opcodes.len(),
    );
    let compat_new = pool
        .rules()
        .iter()
        .filter(|rule| rule.get_name() == "identityel")
        .count();
    let compat_old = pool
        .rules()
        .iter()
        .filter(|rule| rule.get_name() == "identity_el")
        .count();
    println!(
        "lookup=missing|new={}|old={}",
        usize::from(compat_new == 1),
        usize::from(compat_old != 0),
    );
    println!("rule_name={}", pool.rules()[0].get_name());
    println!("desc=x4|stage=before|ops={}", descend_order(&x4, &cases));
    println!("desc=b1|stage=before|ops={}", descend_order(&b1, &cases));
    println!("block|stage=before|ops={}", block_order(&block, &cases));

    let mut state = ActionState::new(rugra::action::action_flags::RULE_REPEATAPPLY);
    let result = pool
        .perform(&mut fd, &mut state)
        .expect("ActionPool perform");
    println!(
        "pool|perform={}|status={}|tests={}|apply={}|rule_tests={}|rule_apply={}",
        result,
        state.status,
        state.count_tests,
        state.count_apply,
        counters.tests.load(Ordering::SeqCst),
        counters.apply.load(Ordering::SeqCst),
    );
    for record in &cases {
        emit_case(record, &block);
    }
    println!("desc=x4|stage=after|ops={}", descend_order(&x4, &cases));
    println!("desc=b1|stage=after|ops={}", descend_order(&b1, &cases));
    println!("block|stage=after|ops={}", block_order(&block, &cases));
}

fn main() {
    run();
}
